sdk/cosmos/azure_data_cosmos/src/location_cache.rs

```diff
@@ -0,0 +1,603 @@
+
+        let endpoint1 = "https://location1.documents.example.com";
+
+        let before_marked_unavailable_time = SystemTime::now();
+        cache.mark_endpoint_unavailable(endpoint1, RequestOperation::Read);
+        cache.mark_endpoint_unavailable(endpoint1, RequestOperation::Write);
+
+        assert!(
+            cache
+                .location_unavailability_info_map
+                .read()
+                .unwrap()
+                .get(endpoint1)
+                .map(|info| info.last_check_time)
+                > Some(before_marked_unavailable_time)
+        );
+
+        assert_eq!(
+            cache
+                .location_unavailability_info_map
+                .read()
+                .unwrap()
+                .get(endpoint1)
+                .map(|info| info.unavailable_operation),
+            Some(RequestOperation::All)
+        );
+    }
+
+    #[test]
+    fn refresh_stale_endpoints() {
+        // create test cache
+        let mut cache = create_test_location_cache();
+
+        // mark endpoint 1 and endpoint 2 as unavailable
+        let endpoint1 = "https://location1.documents.example.com";
+        let endpoint2 = "https://location2.documents.example.com";
+        cache.mark_endpoint_unavailable(endpoint1, RequestOperation::Read);
+        cache.mark_endpoint_unavailable(endpoint2, RequestOperation::Read);
+
+        // simulate stale entry
+        {
+            let mut unavailability_map = cache.location_unavailability_info_map.write().unwrap();
+            if let Some(info) = unavailability_map.get_mut(endpoint1) {
+                info.last_check_time = SystemTime::now() - std::time::Duration::from_secs(500);
+            }
+        }
+
+        // refresh stale endpoints
+        cache.refresh_stale_endpoints();
+
+        // check that endpoint 1 is marked as available again
+        assert!(!cache.is_endpoint_unavailable(endpoint1, RequestOperation::Read));
+    }
+
+    #[test]
+    fn resolve_service_endpoint() {
+        // create test cache
+        let cache = create_test_location_cache();
+
+        // resolve service endpoint
+        let endpoint = cache.resolve_service_endpoint(0, RequestOperation::Read);
+        assert_eq!(
+            endpoint,
+            "https://location1.documents.example.com".to_string()
+        );
+    }
+
+    #[test]
+    fn resolve_service_endpoint_second_location() {
+        // create test cache
+        let cache = create_test_location_cache();
+
+        // resolve service endpoint for second location
+        let endpoint = cache.resolve_service_endpoint(1, RequestOperation::Read);
+        assert_eq!(
+            endpoint,
+            "https://location2.documents.example.com".to_string()
+        );
+    }
+
+    #[test]
+    fn resolve_service_endpoint_default() {
+        let cache = create_test_location_cache();
+
+        let endpoint = cache.resolve_service_endpoint(5, RequestOperation::Read);
```

Comment:
nit: We should be able to access the `read_endpoints` list, so rather than picking a magic number we know is currently more than the number of locations in the test data, we can use the length itself:

```suggestion
        let endpoint = cache.resolve_service_endpoint(cache.read_endpoints.len() + 1, RequestOperation::Read);
```